src/pymor/algorithms/tr.py

```diff
@@ -0,0 +1,306 @@
+    miniter_subproblem
+        Minimum amount of iterations to perform in the BFGS subproblem.
+    maxiter_subproblem
+        Fail the BFGS subproblem if the iteration count reaches this value without converging.
+    tol
+        Finish when the clipped parameter error measure is below this threshold.
+    radius_tol
+        Threshold for increasing the trust region radius upon extending the reduced order model.
+    atol
+        Finish the subproblem when the absolute error measure is below this threshold.
+    tol_sub
+        Finish when the subproblem clipped parameter error measure is below this threshold.
+    line_search_params
+        Dictionary of additional parameters passed to the line search method.
+    stagnation_window
+        Finish when the parameter update has been stagnating within a tolerance of
+        `stagnation_threshold` during the last `stagnation_window` iterations.
+    stagnation_threshold
+        See `stagnation_window`.
+    return_stages
+        If `True`, return a `list` of the intermediate parameter values of `mu` after each
+        trust region iteration.
+    return_subproblem_stages
+        If `True`, return a `list` of the intermediate parameter values of `mu` after each
+        BFGS subproblem iteration.
+
+    Returns
+    -------
+    mu
+        |Numpy array| containing the computed parameter.
+    data
+        Dict containing the following fields:
+
+            :mus:             `list` of parameters after each iteration.
+            :mu_norms:        |NumPy array| of the solution norms after each iteration.
+            :subproblem_data: `list` of data generated by the individual subproblems.
+            :stages:          See `return_stages`.
+
+    Raises
+    ------
+    TRError
+        Raised if the BFGS algorithm failed to converge.
+    """
+    assert shrink_factor != 0.
+
+    logger = getLogger('pymor.algorithms.tr')
+
+    if initial_guess is None:
+        initial_guess = parameter_space.sample_randomly(1)[0]
+        mu = initial_guess.to_numpy()
+    else:
+        mu = initial_guess.to_numpy() if isinstance(initial_guess, Mu) else initial_guess
+
+    surrogate = TRSurrogate(reductor, initial_guess)
+
+    data = {}
+    data['subproblem_data'] = []
+    if return_stages:
+        stages = []
+
+    # compute norms
+    mu_norm = np.linalg.norm(mu)
+    update_norms = []
+    foc_norms = []
+    data['mus'] = [mu.copy()]
+
+    old_rom_output = surrogate.rom_output(mu)
+    old_fom_output = surrogate.fom_output(mu)
+
+    first_order_criticity = 1e6
+    iteration = 0
+    while True:
+        with logger.block(f'Starting adaptive TR algorithm iteration {iteration + 1} with radius {radius}...'):
+            rejected = False
+
+            if iteration >= miniter:
+                if first_order_criticity < tol:
+                    logger.info(
+                        f'TR converged in {iteration} iterations because first order criticity tolerance of {tol}' \
```

Comment:
```suggestion
                        f"TR converged in {iteration} iterations because first order criticity tolerance of {tol}"
```
or give the surrogate a method that returns the basis length.